### PR TITLE
attributes: add new attribute upstream.request_attempt_count

### PR DESCRIPTION
**Pull request by agrawroh**
**Commit Message**: add `upstream.request_attempt_count` to get the number of times a request is attempted upstream
**Additional Description**: Added a new attribute to extract `upstream.request_attempt_count` from the `StreamInfo()`. This could be used in CEL Matcher for access log filtering.
**Risk Level**: Low
**Testing:** Added Unit Tests
**Docs Changes:** Added
**Release Notes:** Added

### CODE REVIEW

File: test/extensions/filters/common/expr/context_test.cc [hunk 1]
```diff
@@ -513,6 +513,8 @@ TEST(Context, ConnectionAttributes) {
       .WillRepeatedly(ReturnRef(connection_termination_details));
   const std::string downstream_transport_failure_reason = "TlsError";
   info.setDownstreamTransportFailureReason(downstream_transport_failure_reason);
+  const int32_t upstream_request_attempt_count = 3;
```
Comment (nezdolik) [inline test/extensions/filters/common/expr/context_test.cc:516]: nit: uint32_t

File: source/extensions/filters/common/expr/context.cc [hunk 2]
```diff
@@ -258,6 +258,8 @@ absl::optional<CelValue> UpstreamWrapper::operator[](CelValue key) const {
   } else if (value == UpstreamTransportFailureReason) {
     return CelValue::CreateStringView(
         info_.upstreamInfo().value().get().upstreamTransportFailureReason());
+  } else if (value == UpstreamRequestAttemptCount) {
+    return CelValue::CreateInt64(info_.attemptCount().value_or(0));
```
Comment (nezdolik) [inline source/extensions/filters/common/expr/context.cc:262]: nit: use `CreateUint64`

File: test/extensions/filters/common/expr/context_test.cc [hunk 3]
```diff
@@ -702,6 +704,13 @@ TEST(Context, ConnectionAttributes) {
     EXPECT_EQ(downstream_transport_failure_reason, value.value().StringOrDie().value());
   }
 
+  {
+    auto value = upstream[CelValue::CreateStringView(UpstreamRequestAttemptCount)];
+    EXPECT_TRUE(value.has_value());
+    ASSERT_TRUE(value.value().IsUint64());
+    EXPECT_EQ(upstream_request_attempt_count, value.value().Int64OrDie());
```
Comment (kyessenov) [inline test/extensions/filters/common/expr/context_test.cc:711]: Uint64OrDie?

File: docs/root/intro/arch_overview/advanced/attributes.rst [hunk 4]
```diff
@@ -152,6 +152,7 @@ The following attributes are available once the upstream connection is establish
    upstream.sha256_peer_certificate_digest, string, SHA256 digest of the peer certificate in the upstream TLS connection if present
    upstream.local_address, string, The local address of the upstream connection
    upstream.transport_failure_reason, string, The upstream transport failure reason e.g. certificate validation failed
+   upstream.request_attempt_count, int, The count of upstream request attempts. A value of ‘0’ indicates that the request was never attempted upstream
```
Comment (kyessenov) [inline docs/root/intro/arch_overview/advanced/attributes.rst:155]: int->uint. In CEL, uint is a distinct type, so you have to write literals as `3u` and convert explicitly if you want uint.

**Review by kyessenov — CHANGES_REQUESTED** [review]
See comments.

File: docs/root/intro/arch_overview/advanced/attributes.rst (hunk 4 above)
Comment (agrawroh) [reply docs/root/intro/arch_overview/advanced/attributes.rst:155]: @kyessenov Thanks for pointing it out. Should we keep this as `uint` or switch it to string?

Comment (kyessenov) [reply docs/root/intro/arch_overview/advanced/attributes.rst]: Don't think it's that important. `int` or `uint` are fine, but please be consistent throughout.

Comment (agrawroh) [reply docs/root/intro/arch_overview/advanced/attributes.rst]: Got it, Thanks. I switched everything to unit

**Review by nezdolik — APPROVED** [review]

**Review by kyessenov — APPROVED** [review]

**Pull request merged**